draft-pauly-taps-arch.md

```diff
@@ -339,6 +339,56 @@ The Transport System Implementation Concepts define the set of objects used inte
 
 * Endpoint Racing: Endpoint Racing is the act of attempting to establish, or scheduling attempts to establish, multiple Protocol Stacks that differ based on the specific representation of the Remote Endpoint and the Local Endpoint, such as IP addresses resolved from a DNS hostname.
 
+### Message Framing, Parsing, and Serialization {#framing}
+
+While some transports expose a byte stream abstraction, most higher level
+protocols impose some structure onto that byte stream. That is, the higher
+level protocol operates in terms of messages, protocol data units (PDUs),
+rather than using unstructured sequences of bytes, with each message being
+processed in turn.  Protocols are specified in terms of state machines
+acting on semantic messages, with parsing the byte stream into messages
+being a necessary annoyance, rather than a semantic concern.  Accordingly,
+the Transport Services architecture exposes messages as the primary 
+abstraction.  Protocols that deal only in byte streams, such as TCP,
+represent their data in each direction as a single, long message.  When
+framing protocols are placed on top of byte streams, the messages used in
+the API represent the framed messages within the stream.
+
+Providing a message-based abstraction also provides:
+
+* the ability to associate deadlines with messages, for transports that
+  care about timing;
+
+*  the ability to provide control of reliability, choosing what messages to
+   retransmit in the event of packet loss, and how best to make use of the
+   data that arrived;
+
+* the ability to manage dependencies between messages, when some messages
+  may not be delivered due to either packet loss or missing a deadline, in
+  particular the ability to avoid (re-)sending data that relies on a previous
+  transmission that was never received.
+
+All require explicit message boundaries, and application-level framing of
+messages, to be effective.  Once a message is passed to the transport, it
+can not be cancelled or paused, but prioritization as well as lifetime and
+retransmission management will provide the protocol stack with all needed
+information to send the messages as quickly as possible without blocking
+transmission unnecessarily. The transport services architecture facilitates
+this by handling messages, with known identity (sequence numbers, in the
+simple case), lifetimes, niceness, and antecedents.
+
+Transport protocols such as SCTP provide a message-oriented API that has
+similar features to those we describe.  Other transports, such as TCP, do
+not.  To support a message oriented API, while still being compatible with
+stream-based transport protocols, implementations of the transport services
+architecture must provide APIs for framing and de-framing messages.  That
```

Comment:
"must" is pretty strong here. I can agree to having this as a "must" on the receiver side for the added value that it can give there, but I'm less sure about the sender side being a "must"

Reply:
I agree that this should be a "should", since a perfectly reasonable use case may not need messages on TCP. There *are* some long stream use cases, like a big file transfer.

Reply:
Well this is getting into the conversation we wanted to have in London, not before, but my concern is "must use a framer if I don't want 50 connections for 50 short messages over TCP".   What's the point of a sender-side framer anyway?  Only that you can offer common functionality in a library - but when I don't care about that, why can't it be good enough to just use "send"?

THAT is my concern about the "must"

Reply:
+1 to should here, though I still think that the "every message starts a new TCP connection" thing is so silly on its face that if someone really, really wants to implement that, it's their own damn fault and we don't have any responsibility at all to save them from themselves.

Sender-side framers are really just a convenience to allow applications to pass down an application-layer object and have it turn into a bytestream with appropriate framing that a receiver-side deframer can deframe (except in the case that the application (1) requires message semantics and (2) provides no framing, which I'd argue means it absolutely can't run over TCP anyway, TAPS or not -- i.e. this is another one of those silly things we have no duty to mention.)

Reply:
Saying “must provide APIs” is not “must use APIs”...

That said, I’m surprised by the resistance to this, since the overwhelming majority of applications have the equivalent of a framer/de-framer in them today as part of their protocol implementation. This isn’t a new feature: it’s breaking out an existing feature to make it reusable.